File: src/Roguelike/Core/Combat/Combatant/Enemy.cs
using Roguelike.Data;
using System;
using System.Collections.Generic;

namespace Roguelike.Core
{
    /// <summary>
    /// Represents an enemy combatant, holding its runtime state and AI logic.
    /// </summary>
    public class Enemy : Combatant
    {
        public EnemyData SourceEnemyData => (EnemyData)SourceData;
        
        private readonly Random rng;
        public Queue<CombatActionData> ActionBucket { get; private set; } = new Queue<CombatActionData>();

        private int _turnsSinceLastSpecial = 999; 

        public Enemy(EnemyData sourceData, Random rng) : base(sourceData)
        {
            this.rng = rng;
            InitializeActionBucket();
        }

        /// <summary>
        /// Fills the action bucket based on the weights in the enemy's data template.
        /// </summary>
        public void InitializeActionBucket()
        {
            ActionBucket.Clear();
            var actionsToShuffle = new List<CombatActionData>();

            foreach (var weightedAction in SourceEnemyData.ActionSet)
            {
                for (int i = 0; i < weightedAction.Weight; i++)
                {
                    actionsToShuffle.Add(weightedAction.Item);
                }
            }
            
            Shuffle(actionsToShuffle);

            foreach (var action in actionsToShuffle)
            {
                ActionBucket.Enqueue(action);
            }
        }
        
        /// <summary>
        /// Retrieves the next action from the bucket, refilling it if necessary.
        /// Respects SpecialAbilityCooldown.
        /// </summary>
        public CombatActionData GetNextAction()
        {
            if (ActionBucket.Count == 0)
            {
                InitializeActionBucket();
            }

            // Try to find a valid action that meets cooldown requirements
            int checks = 0;
            int maxChecks = ActionBucket.Count + 1; // Safety limit

            while (checks < maxChecks)
            {
                var candidate = ActionBucket.Peek();
                bool isSpecial = candidate.Type == ActionType.ApplyStatusEffect || candidate.Type == ActionType.ApplyDeckEffect;
                
                // Check Cooldown: If special and not enough turns passed, skip it
                if (isSpecial && _turnsSinceLastSpecial < SourceEnemyData.SpecialAbilityCooldown)
                {
                    // Move to back
                    ActionBucket.Enqueue(ActionBucket.Dequeue()); 
                    checks++;
                }
                else
                {
                    // Valid action found
                    var action = ActionBucket.Dequeue();
                    if (isSpecial) 
                    {
                        _turnsSinceLastSpecial = 0;
                    }
                    return action;
                }
            }

            // Fallback: If no valid action found (e.g. all special and all on cooldown), just take the next one
            return ActionBucket.Dequeue();
        }

        public void TickCooldowns()
        {
            _turnsSinceLastSpecial++;
        }
        /// <summary>
        /// Returns the upcoming action without consuming it from the bucket.
        /// </summary>
        public CombatActionData PeekNextAction()
        {
            if (ActionBucket.Count == 0)
            {
                InitializeActionBucket();
            }
            return ActionBucket.Peek();
        }

        /// <summary>
        /// Shuffles a list using the Fisher-Yates algorithm.
        /// </summary>
        private void Shuffle(List<CombatActionData> list)
        {
            int n = list.Count;
            while (n > 1)
            {
                n--;
                int k = rng.Next(n + 1);
                var value = list[k];
                list[k] = list[n];
                list[n] = value;
            }
        }
    }
}

File: python_roguelike/core/enemy.py
import random
from collections import deque
from typing import List, Optional
from .combatant import Combatant
from .active_effect import ActiveEffect
from ..data.combatant.enemy_data import EnemyData
from ..data.combat_action_data import CombatActionData
from ..data.enums import ActionType, DecayType
from ..data.status_effect_data import StatusEffectData


class Enemy(Combatant):
    def __init__(self, source_data: EnemyData, rng: random.Random):
        super().__init__(source_data)
        self.source_enemy_data: EnemyData = source_data
        self._rng = rng
        self.action_bucket: deque = deque()
        self._turns_since_last_special: int = 999
        self._initialize_action_bucket()

    def _initialize_action_bucket(self):
        self.action_bucket.clear()
        actions_to_shuffle: List[CombatActionData] = []

        for weighted_action in self.source_enemy_data.action_set:
            for _ in range(weighted_action.weight):
                actions_to_shuffle.append(weighted_action.item)

        self._shuffle(actions_to_shuffle)
        for action in actions_to_shuffle:
            self.action_bucket.append(action)

    def get_next_action(self) -> CombatActionData:
        if not self.action_bucket:
            self._initialize_action_bucket()

        checks = 0
        max_checks = len(self.action_bucket) + 1

        while checks < max_checks:
            candidate = self.action_bucket[0]
            is_special = (
                candidate.type == ActionType.ApplyStatusEffect or
                candidate.type == ActionType.ApplyDeckEffect
            )
            if is_special and self._turns_since_last_special < self.source_enemy_data.special_ability_cooldown:
                # Move to back
                self.action_bucket.append(self.action_bucket.popleft())
                checks += 1
            else:
                action = self.action_bucket.popleft()
                if is_special:
                    self._turns_since_last_special = 0
                return action

        # Fallback
        return self.action_bucket.popleft()

    def peek_next_action(self) -> CombatActionData:
        if not self.action_bucket:
            self._initialize_action_bucket()
        return self.action_bucket[0]

    def tick_cooldowns(self):
        self._turns_since_last_special += 1

    def _shuffle(self, lst: List):
        n = len(lst)
        while n > 1:
            n -= 1
            k = self._rng.randint(0, n)
            lst[k], lst[n] = lst[n], lst[k]
